File: code/amyAssemblyInterpreterCPP/memory.cpp
// The Amy Assembly Programming Language memory manager
// Author: Amy Burnett
// October 29 2021
//========================================================================
// Includes 

#include <iostream>
#include <vector>
#include <unordered_map>

#include "memory.hpp"

//========================================================================

BlockHeader::BlockHeader ()
{
    prevBlock = nullptr;
    payloadSize = 0;
    isAlloc = false;
    address = 0;
}

//========================================================================

BlockHeader::BlockHeader (BlockHeader* _prevBlock, int _payloadSize, bool _isAlloc, int _address)
{
    prevBlock = _prevBlock;
    payloadSize = _payloadSize;
    isAlloc = _isAlloc;
    address = _address;
}

//========================================================================

Heap::Heap ()
{
    memory = std::vector<Data> (INITIAL_CAPACITY);
    // set up first free block 
    blockHeaders[1] = BlockHeader (nullptr, INITIAL_CAPACITY-2, false, 1);
    // mem[0] is reserved for NULLPTR
}

//========================================================================
// allocates space for given size and returns the pointer to 
//    the starting position

int 
Heap::malloc (int size)
{
    // Search for sufficient mem block
    int i = 1;
    BlockHeader* header = nullptr; 
    while (i < memory.size())
    {
        // get header
        header = &blockHeaders[i];
        // ensure it is free
        if (header->isAlloc)
        {
            i += header->payloadSize + 1;
            continue;
        }
        // if sufficient 
        if (header->payloadSize == size)
        {
            // allocate 
            header->isAlloc = true; 
            // return pointer to first payload pos
            return i + 1;
        }
        // if sufficient
        if (header->payloadSize > size)
        {
            // split payload
            int secondPayload = header->payloadSize - size - 1;
            int secondAddress = i+size+1;
            // modify this payload
            header->payloadSize = size;
            header->isAlloc = true; 
            // if second split has >0 payload
            // *** removed if because caused error 
            // if secondPayload > 0:
                // create new payload
            // create new payload
            blockHeaders[secondAddress] = BlockHeader(header, secondPayload, false, secondAddress);
            memory[secondAddress] = Data('x');
            // assign second split's next's prev to the second split
            if (secondAddress+secondPayload+1 < memory.size())
                blockHeaders[secondAddress+secondPayload+1].prevBlock = &blockHeaders[secondAddress];
            // coalesce 
            coalesce(secondAddress);
            // ****
            // return pointer to first payload pos
            return i + 1;
        }
        i += header->payloadSize + 1;
    }
    // Reachs here if no sufficient free blocks were found
    // Lets add one
    memory.resize (memory.size() + size+1); // plus one for header
    blockHeaders[i] = BlockHeader(header, size, true, i);
    memory[i] = Data('x');
    // return pointer to first payload pos
    return i + 1;
}

//========================================================================
// Frees the block containing the address

int 
Heap::free (int address)
{
    blockHeaders[address-1].isAlloc = false;
    coalesce (address-1);
}

//========================================================================
// Merges the free block at given address with
//   adjacent free blocks

int 
Heap::coalesce (int address)
{
    // combine with free block after
    int nextBlockAddr = address + blockHeaders[address].payloadSize + 1;
    if (nextBlockAddr < memory.size() && !blockHeaders[nextBlockAddr].isAlloc)
    {
        // clear second block
        int temp = blockHeaders[nextBlockAddr].payloadSize;
        blockHeaders.erase (nextBlockAddr);
        memory[nextBlockAddr].i = 0;
        // add space to this block
        // + 1 for the space that used to have the header
        blockHeaders[address].payloadSize += temp+1;
        // assign this to the prev of the next header
        int newNextBlockAddr = address+blockHeaders[address].payloadSize+1; // +1 for this blockheader
        if (newNextBlockAddr < memory.size())
            blockHeaders[newNextBlockAddr].prevBlock = &blockHeaders[address];
    }
    // Combine with free block before
    if (blockHeaders[address].prevBlock != nullptr && !blockHeaders[address].prevBlock->isAlloc)
    {
        // add this space to the prev block payload
        blockHeaders[address].prevBlock->payloadSize += 1 + blockHeaders[address].payloadSize;
        // assign this next to the prev
        // if there is a next
        if (address + blockHeaders[address].payloadSize + 1 < memory.size())
            blockHeaders[address+blockHeaders[address].payloadSize+1].prevBlock = blockHeaders[address].prevBlock;
        // delete header
        blockHeaders.erase (address);
        memory[address].i = 0;
    }
}

//========================================================================
// Returns the payload size for a given pointer
// Pre-condition: pointer must be allocated and not offset 
//   from the start of the block
int 
Heap::sizeOf (int address)
{
    return blockHeaders[address-1].payloadSize;
}

//========================================================================

void
printheap (Heap heap)
{
    for (int i = 0; i < heap.memory.size(); ++i)
    {
        // check if i is a block
        if (heap.blockHeaders.find(i) != heap.blockHeaders.end())
            std::cout << i << " -> " << heap.blockHeaders[i].payloadSize << " " << heap.blockHeaders[i].isAlloc << std::endl;
        else
            std::cout << i << " " << heap.memory[i].toString() << std::endl;
    }
}

//========================================================================

void
sampleRun ()
{
    Heap heap; 

    printheap(heap);
    std::cout << "======================" << std::endl;
    std::cout << "Malloc 6" << std::endl;
    int a = heap.malloc(6);
    printheap(heap);
    std::cout << "pointer -> " << a << std::endl;
    std::cout << "======================" << std::endl;
    std::cout << "Malloc 3" << std::endl;
    int b = heap.malloc(3);
    printheap(heap);
    std::cout << "pointer -> " << b << std::endl;
    std::cout << "======================" << std::endl;
    std::cout << "Malloc 5" << std::endl;
    int c = heap.malloc(5);
    printheap(heap);
    std::cout << "pointer -> " << c << std::endl;
    heap.memory[c+0] = 'h';
    heap.memory[c+1] = 'e';
    heap.memory[c+2] = 'l';
    heap.memory[c+3] = 'l';
    heap.memory[c+4] = 'o';
    std::cout << "======================" << std::endl;
    std::cout << "Free A" << std::endl;
    heap.free(a);
    printheap(heap);
    std::cout << "======================" << std::endl;
    std::cout << "Malloc 3" << std::endl;
    int d = heap.malloc(3);
    printheap(heap);
    std::cout << "pointer -> " << d << std::endl;
    std::cout << "======================" << std::endl;
    std::cout << "Free D" << std::endl;
    heap.free(d);
    printheap(heap);
    std::cout << "======================" << std::endl;
    std::cout << "Free B" << std::endl;
    heap.free(b);
    printheap(heap);
    std::cout << "======================" << std::endl;
    std::cout << "Malloc 3" << std::endl;
    a = heap.malloc(3);
    printheap(heap);
    std::cout << "pointer -> " << a << std::endl;
    std::cout << "======================" << std::endl;
    std::cout << "Malloc 6" << std::endl;
    int e = heap.malloc(6);
    printheap(heap);
    std::cout << "pointer -> " << e << std::endl;
    std::cout << "======================" << std::endl;
    std::cout << "Free A" << std::endl;
    heap.free(a);
    printheap(heap);
    std::cout << "======================" << std::endl;
    std::cout << "Free C" << std::endl;
    heap.free(c);
    printheap(heap);
    std::cout << "======================" << std::endl;
    std::cout << "Free e" << std::endl;
    heap.free(e);
    printheap(heap);
    std::cout << "======================" << std::endl;
}





   

 
   

File: code/amyasmi/memory.py
# The Amy Programming Language Memory Heap
# By Amy Burnett
# November 5 2020
##########################################################################
# Imports

##########################################################################

MEMORY_NULL = 0

class BlockHeader:
    def __init__(self, prevBlock, payloadSize:int, isAlloc:bool):
        self.prevBlock = prevBlock
        self.payloadSize = payloadSize
        self.isAlloc = isAlloc
    def __str__ (self):
        return (f"[{self.payloadSize}, {self.isAlloc}]")
    def __repr__ (self):
        return self.__str__ ()

class Heap:
    def __init__(self):
        initialSize = 100000
        self.memory = [0 for _ in range(initialSize)]
        # set up first free block
        # mem[0] is reserved for NULLPTR
        self.memory[1] = BlockHeader(None, initialSize-2, False)
        self.allocatedMemory = 0

    def malloc(self, size:int) -> int:
        """allocates space for given size and returns the pointer to 
        the starting position"""
        # Search for sufficient mem block
        i = 1
        header = None
        while i < len(self.memory):
            # get header
            header = self.memory[i]
            # ensure it is free
            if header.isAlloc:
                i += header.payloadSize + 1
                continue
            # if sufficient 
            if header.payloadSize == size:
                # allocate this block
                header.isAlloc = True
                self.allocatedMemory += header.payloadSize
                # return pointer to first payload pos
                return i + 1
            # if sufficient
            if header.payloadSize > size:
                # split payload
                secondPayload = header.payloadSize - size - 1
                secondAddress = i+size+1
                # modify this payload
                header.payloadSize = size
                header.isAlloc = True
                self.allocatedMemory += header.payloadSize
                # if second split has >0 payload
                # *** removed if because caused error 
                # if secondPayload > 0:
                    # create new payload
                self.memory[secondAddress] = BlockHeader(header, secondPayload, False)
                # coalesce 
                self.coallese(secondAddress)
                # assign second split's next's prev to the second split
                if secondAddress+secondPayload+1 < len(self.memory):
                    self.memory[secondAddress+secondPayload+1].prevBlock = self.memory[secondAddress]
                # ****
                # return pointer to first payload pos
                return i + 1
            i += header.payloadSize + 1
        # Reachs here if no sufficient free blocks were found
        # Lets add one
        self.memory += [BlockHeader(header, size, True)]
        self.memory += [0 for _ in range(size)]
        self.allocatedMemory += size
        # return pointer to first payload pos
        return i + 1

    def free(self, address:int):
        """Frees the block containing the address"""
        self.memory[address-1].isAlloc = False
        self.allocatedMemory -= self.memory[address-1].payloadSize
        self.coallese(address-1)
    
    def coallese(self, address:int):
        """
        Merges the free block at given address with adjacent free blocks
        - allocated blocks are left unchanged
        """
        # combine with free block after
        nextBlockAddr = address+self.memory[address].payloadSize+1
        if nextBlockAddr < len(self.memory) and not self.memory[nextBlockAddr].isAlloc:
            # clear second split
            temp = self.memory[nextBlockAddr].payloadSize
            self.memory[nextBlockAddr] = 0
            # add space to this block
            # + 1 for the space that used to have the header
            self.memory[address].payloadSize += temp+1
            # assign this to the prev of the next header
            newNextBlockAddr = address+self.memory[address].payloadSize+1
            if newNextBlockAddr < len(self.memory):
                self.memory[newNextBlockAddr].prevBlock = self.memory[address]
        # Combine with free block before
        if self.memory[address].prevBlock != None and not self.memory[address].prevBlock.isAlloc:
            # add this space to the prev block payload
            self.memory[address].prevBlock.payloadSize += 1+self.memory[address].payloadSize
            # assign this next to the prev
            # if there is a next
            if address+self.memory[address].payloadSize+1 < len(self.memory):
                self.memory[address+self.memory[address].payloadSize+1].prevBlock = self.memory[address].prevBlock
            # delete header
            self.memory[address] = 0

    def sizeof(self, address):
        """Returns the payload size for a given pointer

        Pre-condition: pointer must be allocated and not offset 
        from the start of the block
        """
        return self.memory[address-1].payloadSize




def printheap(heap):
    for i in range(len(heap.memory)):
        if isinstance(heap.memory[i], BlockHeader):
            print(f"{i} {heap.memory[i].payloadSize} {heap.memory[i].isAlloc}")
            # if heap.memory[i].prevBlock != None:
            #     print(f"\t{heap.memory[i].prevBlock.payloadSize} {heap.memory[i].prevBlock.isAlloc}")
        else:
            print(f"{i} {heap.memory[i]}")


def sampleRun():
    heap = Heap()

    printheap(heap)
    print("======================")
    print("Malloc 6")
    a = heap.malloc(6)
    printheap(heap)
    print("pointer -> ", a)
    print("======================")
    print("Malloc 3")
    b = heap.malloc(3)
    printheap(heap)
    print("pointer -> ", b)
    print("======================")
    print("Malloc 5")
    c = heap.malloc(5)
    printheap(heap)
    print("pointer -> ", c)
    print("======================")
    print("Free A")
    heap.free(a)
    printheap(heap)
    print("======================")
    print("Malloc 3")
    d = heap.malloc(3)
    printheap(heap)
    print("pointer -> ", d)
    print("======================")
    print("Free D")
    heap.free(d)
    printheap(heap)
    print("======================")
    print("Free B")
    heap.free(b)
    printheap(heap)
    print("======================")
    print("Malloc 3")
    a = heap.malloc(3)
    printheap(heap)
    print("pointer -> ", a)
    print("======================")
    print("Malloc 6")
    e = heap.malloc(6)
    printheap(heap)
    print("pointer -> ", e)
    print("======================")
    print("Free A")
    heap.free(a)
    printheap(heap)
    print("======================")
    print("Free C")
    heap.free(c)
    printheap(heap)
    print("======================")
    print("Free e")
    heap.free(e)
    printheap(heap)
    print("======================")
